src/Hosting/Hosting/src/Internal/HostingMetrics.cs

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Metrics;
+
+namespace Microsoft.AspNetCore.Hosting;
+
+internal sealed class HostingMetrics : IDisposable
+{
+    public const string MeterName = "Microsoft.AspNetCore.Hosting";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentRequestsCounter;
+    private readonly Histogram<double> _requestDuration;
+
+    public HostingMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentRequestsCounter = _meter.CreateUpDownCounter<long>(
+            "current-requests",
+            description: "Current requests");
+
+        _requestDuration = _meter.CreateHistogram<double>(
+            "request-duration",
+            description: "Request duration");
+    }
+
+    public void RequestStart(string scheme, string method, HostString host)
+    {
+        // Tags must match request end.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(1, tags);
+    }
+
+    public void RequestEnd(string scheme, string method, HostString host, string? routePattern, int statusCode, Exception? exception, IList<KeyValuePair<string, object?>>? customTags, long startTimestamp, long currentTimestamp)
+    {
+        // Tags must match request start.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(-1, tags);
+
+        // Add information gathered during request.
+        tags.Add("status-code", GetBoxedStatusCode(statusCode));
+        if (routePattern != null)
+        {
+            tags.Add("route", routePattern);
+        }
+        // This exception is only present if there is an unhandled exception.
+        // Exceptions caught by ExceptionHandlerMiddleware and DeveloperExceptionMiddleware add exception-name to custom tags.
+        if (exception != null)
+        {
+            tags.Add("exception-name", exception.GetType().Name);
+        }
+        if (customTags != null)
+        {
+            for (var i = 0; i < customTags.Count; i++)
+            {
+                tags.Add(customTags[i]);
```

Comment:
How are the naming conflicts handled? if a tag with a name that already exists e.g. `host` or `method`.

Another thing is, while customTags allows you to provide additional tags to be added here, how can one override value of an existing tag. E.g. consider this case

You are adding `route` if present, but the user always wants `route` to be present because that's critical for their metrics alerting and insights. For cases where route isn't available from routePattern, user might have additional logic in their code to provide a route. Or a customer might want to use a friendly name for a request instead of the route always because that's more easy to map the request to their code when looking at the metrics in their dashboards and alerts.

Reply:
The measurement will have two tags with the same name if duplicates exist.

In the example of route, that can be customized by specifying metadata on the endpoint.

Reply:
Duplicate tags aren't supported with metrics AFAIK. @reyang to confirm if OpenTelemetry supports it and what's the behavior when a metric has duplicate tag?

Reply:
There's an extension point to customize metadata on an endpoint today. It's possible to override the template route with a custom `IRouteDiagnosticsMetadata` value.

The custom tags are added to the `TagList` last. If there is a "last wins" policy by whoever uses the measurements, then the custom tag value would take precedence. But that's up to whoever uses the measurement. @noahfalk Are you aware of anything in the metrics spec about this situation?

Adding a note to metrics docs that you shouldn't add custom tags that conflict with a known tag name for a counter seems sensible.

Reply:
> Are you aware of anything in the metrics spec about this situation?

The spec says the attributes in a collection should always be uniquely named but leaves it up to implementations as to how that is enforced. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute-collections

.NET TagList+Meter+MeterListener do not inspect or enforce anything about the tag content. The tags will passed through exactly with the same ordering and naming that they were provided in.

Currently I don't think MetricsEventSource is enforcing any duplication rules and you probably see duplicate attributes in the output with unspecified ordering from input to output. For example if you made these calls:

counter.Add(key1="apple",route="foo",route="bar", Measurement=10)
counter.Add(route="bar",route="foo",key1="apple", Measurement=20)

MetricsEventSource makes no promises whether it treats those as the same set of keys and you get a single time series with value=30 or it is two separate time series with values 10 and 20. If the two route attributes had distinct names (route1=foo,route2=bar) then you would be guaranteed to get a single time series with value 30 regardless of the input order.

I don't know offhand if OpenTelemetry makes any stronger promises. I would guess at best it would reliably give precedence based on order and at worst it is similar to MetricsEventSource giving undefined behavior.

I doubt ordering is something we'd want users of the API to rely on given that .NET isn't in position to enforce how different listeners will choose to handle duplicate key names.